File: Python/Types of Matrix/Idempotent matrix.py
def matrixMultiply(A, B):
    result = [[None]*len(B[0]) for row in range(len(A))]
    for row in range(0, len(A)):
        for col in range(0, len(B[row])):
            calc = 0
            for m in range(0, len(B)):
                calc += A[row][m] * B[m][col]
            result[row][col] = calc
    return result


# Idempotent matrix is a square matrix which when multiplied by itself, gives back the same matrix (A^2 = A)
A = [[ 2,  1],
     [-2, -1]]

if len(A) == len(A[0]):
    print("Matrix: ")
    for row in range(0, len(A)):
        for col in range(0, len(A[row])):
            print("\t", A[row][col], end='')
        print()
    
    result = matrixMultiply(A, A)
    # Print idempotent matrix after calculation
    print("\nIdempotent matrix: ")
    for row in range(0, len(result)):
        for col in range(0, len(result[row])):
            print("\t", result[row][col], end='')
        print()
else:
    print("Error! Multiplication not possible")

File: C++/Types of Matrix/Idempotent matrix.cpp
#include <iostream>
#include <vector>
using namespace std;

vector <vector<int>> matrixMultiply(vector <vector<int>> A, vector <vector<int>> B) {
    vector <vector<int>> result (A.size(), vector <int> (B[0].size()));
    int calc;
    for (int row = 0; row < A.size(); row++) {
        for (int col = 0; col < B[row].size(); col++) {
            calc = 0;
            for (int m = 0; m < B.size(); m++) {
                calc += A[row][m] * B[m][col];
            }
            result[row][col] = calc;
        }
    }
    return result;
}

int main(){
    // Idempotent matrix is a square matrix which when multiplied by itself, gives back the same matrix (A^2 = A)
    vector <vector<int>> A = {
        { 2,  1 },
        {-2, -1 }
    };

    if(A.size() == A[0].size()){
        cout << "Matrix: " << endl;
        for (int row = 0; row < A.size(); row++) {
            for (int col = 0; col < A[row].size(); col++) {
                cout << "\t" << A[row][col];
            }
            cout << endl;
        }
        
        vector <vector<int>> result = matrixMultiply(A, A);
        // Print idempotent matrix after calculation
        cout <<"\nIdempotent matrix: " << endl;
        for (int row = 0; row < result.size(); row++) {
            for (int col = 0; col < result[row].size(); col++) {
                cout << "\t" << result[row][col];
            }
            cout << endl;
        }
    }else{
        cout << "Error! Multiplication not possible" << endl;
    }
}